gas_price.go

```diff
@@ -29,44 +38,79 @@ const (
 	GasPriorityAverage = GasPriority("average")
 )
 
-// SuggestGasPrice returns a suggested gas price value in wei (base units) for timely transaction execution.
+// SuggestGasPrice returns a suggested gas price value in wei (base units) for timely transaction execution. It always
+// makes a new call to the ETH Gas Station API. Use NewGasPriceSuggester to leverage cached results.
 //
 // The returned price depends on the priority specified, and supports all priorities supported by the ETH Gas Station API.
 func SuggestGasPrice(priority GasPriority) (*big.Int, error) {
 	prices, err := loadGasPrices()
 	if err != nil {
 		return nil, err
 	}
-
-	switch priority {
-	case GasPriorityFast:
-		return parseGasPriceToWei(prices.Fast)
-	case GasPriorityFastest:
-		return parseGasPriceToWei(prices.Fastest)
-	case GasPrioritySafeLow:
-		return parseGasPriceToWei(prices.SafeLow)
-	case GasPriorityAverage:
-		return parseGasPriceToWei(prices.Average)
-	default:
-		return nil, errors.New("eth: unknown/unsupported gas priority")
-	}
+	return parseSuggestedGasPrice(priority, prices)
 }
 
 // SuggestFastGasPrice is a helper method that calls SuggestGasPrice with GasPriorityFast
+//
+// It always makes a new call to the ETH Gas Station API. Use NewGasPriceSuggester to leverage cached results.
 func SuggestFastGasPrice() (*big.Int, error) {
 	return SuggestGasPrice(GasPriorityFast)
 }
 
+// NewGasPriceSuggester returns a function that can be used to either load a new gas price response, or use a cached
+// response if it is within the age range defined by maxResultAge.
+//
+// The returned function loads from the cache or pulls a new response if the stored result is older than maxResultAge.
+func NewGasPriceSuggester(maxResultAge time.Duration) (GasPriceSuggester, error) {
+	prices, err := loadGasPrices()
+	if err != nil {
+		return nil, err
+	}
+	m := gasPriceManager{
+		latestResponse: prices,
+		fetchedAt:      time.Now(),
+		maxResultAge:   maxResultAge,
+	}
+	return func(priority GasPriority) (*big.Int, error) {
+		return m.suggestCachedGasPrice(priority)
+	}, nil
+}
+
+type gasPriceManager struct {
+	latestResponse *ethGasStationResponse
+	fetchedAt      time.Time
+	maxResultAge   time.Duration
+
+	mux sync.Mutex
```

Comment:
There's a cool feature of Go that I discovered recently which allows a struct to have `Lock` and `Unlock` methods. If you leave off the field name of `mux`, then you can just use `m.Lock()` and `m.Unlock()`. It's pretty slick when you only have one mutex

Reply:
Ahh I like that! I'll embed the mutex in that case